File: 14-TypeScript/Grey Multi-Tenant/packages/grey-r/R/query_client.R
#' @title Query Client
#' @description Query client for executing read operations
#' @name QueryClient
NULL

#' @importFrom R6 R6Class
#' @export
QueryClient <- R6::R6Class(
  "QueryClient",
  
  public = list(
    #' @description Create a new query client
    #' @param http_client HttpClient instance
    initialize = function(http_client) {
      private$http <- http_client
    },
    
    #' @description Execute a query operation
    #' @param query_name The name of the query
    #' @param variables Query variables (optional)
    #' @return grey_result with query result or error
    query = function(query_name, variables = NULL) {
      if (is.null(query_name) || nchar(trimws(query_name)) == 0) {
        return(result_err(error_validation("Query name is required")))
      }
      
      body <- list(query = query_name)
      if (!is.null(variables)) body$variables <- variables
      
      private$http$post("/api/v1/query", body)
    },
    
    #' @description Execute a batch of queries
    #' @param queries List of query objects with query_name and variables
    #' @return grey_result with array of results or error
    batch_query = function(queries) {
      if (is.null(queries) || !is.list(queries)) {
        return(result_err(error_validation("Queries list is required")))
      }
      
      if (length(queries) == 0) {
        return(result_err(error_validation("At least one query is required")))
      }
      
      formatted <- lapply(queries, function(q) {
        list(
          query = q$query_name,
          variables = q$variables
        )
      })
      
      private$http$post("/api/v1/query/batch", list(queries = formatted))
    },
    
    #' @description Execute a paginated query
    #' @param query_name The query name
    #' @param variables Query variables (optional)
    #' @param page Page number (default: 1)
    #' @param per_page Items per page (default: 20)
    #' @return grey_result with paginated result or error
    paginated_query = function(query_name, variables = NULL, page = 1, per_page = 20) {
      if (is.null(query_name) || nchar(trimws(query_name)) == 0) {
        return(result_err(error_validation("Query name is required")))
      }
      
      body <- list(
        query = query_name,
        variables = variables %||% list(),
        pagination = list(
          page = page,
          per_page = per_page
        )
      )
      
      private$http$post("/api/v1/query", body)
    }
  ),
  
  private = list(
    http = NULL
  )
)

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-ruby/lib/grey_sdk/domain/query_client.rb
# frozen_string_literal: true

module GreySdk
  module Domain
    # Query client for executing read operations
    class QueryClient
      # @param http_client [GreySdk::Http::HttpClient] The HTTP client
      def initialize(http_client)
        @http = http_client
      end

      # Executes a query operation
      # @param query_name [String] The name of the query to execute
      # @param variables [Hash, nil] Query variables
      # @return [GreySdk::Error::Result] Result containing query result or error
      def query(query_name:, variables: nil)
        return validation_error("Query name is required") if query_name.nil? || query_name.empty?

        body = { query: query_name }
        body[:variables] = variables if variables

        @http.post("/api/v1/query", body)
      end

      # Executes a batch of queries
      # @param queries [Array<Hash>] Array of query objects with :query_name and :variables
      # @return [GreySdk::Error::Result] Result containing array of results or error
      def batch_query(queries:)
        return validation_error("Queries array is required") if queries.nil? || !queries.is_a?(Array)
        return validation_error("At least one query is required") if queries.empty?

        body = {
          queries: queries.map do |q|
            {
              query: q[:query_name],
              variables: q[:variables]
            }
          end
        }

        @http.post("/api/v1/query/batch", body)
      end

      # Executes a named query with pagination
      # @param query_name [String] The name of the query
      # @param variables [Hash, nil] Query variables
      # @param page [Integer] Page number
      # @param per_page [Integer] Items per page
      # @return [GreySdk::Error::Result] Result containing paginated result or error
      def paginated_query(query_name:, variables: nil, page: 1, per_page: 20)
        return validation_error("Query name is required") if query_name.nil? || query_name.empty?

        body = {
          query: query_name,
          variables: variables || {},
          pagination: {
            page: page,
            per_page: per_page
          }
        }

        @http.post("/api/v1/query", body)
      end

      private

      def validation_error(message)
        Error::Result.err(Error::GreyError.validation(message))
      end
    end
  end
end
